crates/wasmtime/src/instance.rs

```diff
@@ -658,16 +661,33 @@ pub struct InstancePre<T> {
     /// preallocate space in a `Store` up front for all entries to be inserted.
     host_funcs: usize,
 
+    /// The `VMCallerCheckedFuncRef`s for the functions in `items` that do not
+    /// have a `wasm_call` trampoline. We pre-allocate and pre-patch these
+    /// `VMCallerCheckedFuncRef`s so that we don't have to do it at
+    /// instantiation time.
+    ///
+    /// This is an `Arc<[T]>` for the same reason as `items`.
+    func_refs: Arc<[VMCallerCheckedFuncRef]>,
+
     _marker: std::marker::PhantomData<fn() -> T>,
 }
 
+// Safety: We need this manual implementations because of
+// `trampolines_for_patching` and its `NonNull` pointers, which are not
+// send/sync by default. However, they are logically part of the `Module` and
+// are safe to send and access across threads (provided you work out the
+// appropriate lifetimes, of course).
+unsafe impl<T> Send for InstancePre<T> {}
+unsafe impl<T> Sync for InstancePre<T> {}
```

Comment:
Like with `FuncData`, this feels a bit too broad to me. Could these impls be placed on `VMCallerCheckedFuncRef` instead?